### PR TITLE
Support ASGI3

### PR DESCRIPTION
**Pull request by tomchristie**
Refs https://github.com/django/asgiref/pull/80

Notes:

* I've not yet added tests here - not sure where to do so.
* I figure it makes more sense to go with the more minimal "middleware shim if it's ASGI3", rather than switching the default to ASGI3, and using "middleware shim if it's ASGI2". A follow-up could switch that around at a later date.
* I've taken the auto-detect directly from asgiref. Uvicorn uses a [slightly different approach](https://github.com/encode/uvicorn/blob/master/uvicorn/config.py#L185-L191).

### CODE REVIEW

File: daphne/cli.py [hunk 1]
```diff
@@ -14,6 +17,17 @@
 DEFAULT_PORT = 8000
 
 
+class ASGI3Middleware:
+    def __init__(self, app):
+        self.app = app
+
+    def __call__(self, scope):
+        return functools.partial(self.asgi, scope=scope)
```
Comment (pgjones) [inline daphne/cli.py]: @tomchristie I think the middleware should change the asgi version in the scope.

Comment (tomchristie) [reply daphne/cli.py]: Good call.

**Comment by carltongibson** [comment]
Hey Tom, thanks super. I'll ask @andrewgodwin to review.

File: daphne/cli.py [hunk 2]
```diff
@@ -198,6 +221,23 @@ def _get_forwarded_port(self, args: Namespace):
         if args.proxy_headers:
             return "X-Forwarded-Port"
 
+    def _guess_asgi_protocol(self, application):
```
Comment (andrewgodwin) [inline daphne/cli.py]: Why not just use `asgiref.compatability` for this once we merge it? I'd rather not have it reimplemented everywhere.

**Comment by tomchristie** [comment]
Yup, that’d be grand. I’ll do that.

**Comment by andrewgodwin** [comment]
Maybe use `asgiref~=3.0`, but other than that, looks good.

**Comment by tomchristie** [comment]
Gotcha 👍

**Comment by tomchristie** [comment]
Reckon we could get a daphne release with this?

I'm super-keen on getting hypercorn and daphne releases with ASGI 3 support baked in, since it's starting to block a few things in the ecosystem. It'd also be really valuable for me to be able to go into DjangoCon Europe (this coming week) and unambiguously point at ASGI 3 as being *"this is how it works"*.

**Comment by carltongibson** [comment]
Hey Tom. Yes. 

It’s my target for the sprints at DjangoCon next week to go through this issue and related and do releases and such. 

However if you need it before then... 
(Let me check I’ve got the right PyPI powers...)

**Comment by carltongibson** [comment]
Ok, right... no I don't. @andrewgodwin I'm carltongibson on PyPI https://pypi.org/user/carltongibson/. If you give me the powers I'm happy to roll a release here. (Otherwise you have to do it. 😀)

**Comment by andrewgodwin** [comment]
Actually Travis can auto-release for you if you push a tag; see here: https://channels.readthedocs.io/en/latest/contributing.html#how-do-i-do-a-release

I'll go give you actual release power too, though.

**Comment by carltongibson** [comment]
Brilliant. (“Read the fine manual”). Even better. 😀

OK. I’ll sort this out before the conf. 👍

**Comment by carltongibson** [comment]
Yep. Need to bump the asgiref version.

**Pull request merged**